File: LeetCode/0059-spiral-matrix-ii/solution.py
from typing import List


class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        result = [[0 for _ in range(n)] for _ in range(n)]
        offset = 0  # 偏移量
        num = 1  # 需要填充的数字
        loop = n // 2  # 一共需要画的轮数
        step = n - 1  # 画每条边需要的次数

        for _ in range(loop):

            for x in range(step):

                result[0+offset][x+offset] = num
                num += 1

            for y in range(step):

                result[y+offset][n-1-offset] = num
                num += 1

            for x in range(step, 0, -1):

                result[n-1-offset][x+offset] = num
                num += 1

            for y in range(step, 0, -1):

                result[y+offset][0+offset] = num
                num += 1

            offset += 1
            step -= 2

        # 如果 n 为奇数，需要单独给中心点赋值
        if (n % 2 == 1):
            mid = int((n - 1) / 2)
            result[mid][mid] = num

        return result

File: LeetCode/0059-spiral-matrix-ii/Solution.java
/**
 * 59. 螺旋矩阵 II 左闭右开画法
 */
class Solution {
    public int[][] generateMatrix(int n) {
        int[][] matrix = new int[n][n];

        int startX = 0;
        int startY = 0;
        int loop = n / 2;
        int count = 1;
        int offset = 1;
        int x;
        int y;

        while (loop-- > 0) {
            // 填充上边
            for (x = startX; x < n - offset; x++) {
                matrix[startY][x] = count++;
            }
            // 填充右边
            for (y = startY; y < n - offset; y++) {
                matrix[y][x] = count++;
            }
            // 填充下边
            for (; x > startX; x--) {
                matrix[y][x] = count++;
            }
            // 填充左边
            for (; y > startY; y--) {
                matrix[y][x] = count++;
            }

            startX++;
            startY++;
            offset++;
        }

        if (n % 2 == 1) {
            int mid = n / 2;
            matrix[mid][mid] = count;
        }

        return matrix;
    }
}
